src/planner/binder/query_node/bind_select_node.cpp

```diff
@@ -445,15 +460,25 @@ unique_ptr<BoundQueryNode> Binder::BindSelectNode(SelectNode &statement, unique_
 		bool is_original_column = i < result->column_count;
 		bool can_group_by_all =
 		    statement.aggregate_handling == AggregateHandling::FORCE_AGGREGATES && is_original_column;
+
+		// we still need to push back a dummy result, as we might bound a modifier (ORDER BY) previously,
+		// which will otherwise have a wrong column index in BindModifierTypes
+		modifier_sql_types.push_back(result_type);
+		auto prev_count = modifier_expansion_count.empty() ? 0 : modifier_expansion_count.back();
+		modifier_expansion_count.emplace_back(prev_count);
+
 		if (select_binder.HasExpandedExpressions()) {
 			if (!is_original_column) {
 				throw InternalException("Only original columns can have expanded expressions");
 			}
 			if (statement.aggregate_handling == AggregateHandling::FORCE_AGGREGATES) {
 				throw BinderException("UNNEST of struct cannot be combined with GROUP BY ALL");
 			}
+
 			auto &struct_expressions = select_binder.ExpandedExpressions();
 			D_ASSERT(!struct_expressions.empty());
+			modifier_expansion_count.back() += struct_expressions.size() - 1;
```

Comment:
Couldn't we instead also push the types of the struct expansion into `internal_sql_types` here so they are available while binding the ORDER BY?

Reply:
I tried that, but we still don't know how much the column index of the ORDER BY column changes due to the expansion, so we can't get the correct type later if we don't know the shift (I think). `modifier_expansion_count` is to keep track of that. And to update the column index correctly. If I follow the code correctly, we first bind the ORDER BY modifiers, so their column indexes don't account for any potential expansions.

Reply:
But I understand the concern of pushing the dummy type. I will look more into it.

Reply:
Perhaps we could do the expansion in an earlier stage instead, which is what we do for e.g. `*` expressions?